cmake/common_compiler_flags.cmake

```diff
@@ -1,94 +1,162 @@
-    >
-    $<$<AND:${compiler_is_gnu},${compiler_greater_than_or_equal_v9}>:
-        -Wattribute-alias=2
-    >
-    $<$<AND:${compiler_is_gnu},${compiler_greater_than_or_equal_v11}>:
-        # Broke on MethodBind templates before GCC 11.
-        -Wlogical-op
-    >
-    $<$<AND:${compiler_is_gnu},${compiler_less_than_v11}>:
-        # Regression in GCC 9/10, spams so much in our variadic templates that we need to outright disable it.
-        -Wno-type-limits
-    >
-    $<$<AND:${compiler_is_gnu},${compiler_greater_than_or_equal_v12}>:
-        # False positives in our error macros, see GH-58747.
-        -Wno-return-type
-    >
-)
-
-# Treat warnings as errors
-function( set_warning_as_error )
-    message( STATUS "[${PROJECT_NAME}] Treating warnings as errors")
-    if ( CMAKE_VERSION VERSION_GREATER_EQUAL "3.24" )
-        set_target_properties( ${PROJECT_NAME}
-            PROPERTIES
-                COMPILE_WARNING_AS_ERROR ON
-        )
-    else()
-        target_compile_options( ${PROJECT_NAME}
-            PRIVATE
-                $<${compiler_is_msvc}:/WX>
-                $<$<OR:${compiler_is_clang},${compiler_is_gnu}>:-Werror>
-        )
-    endif()
-endfunction()
+#[=======================================================================[.rst:
+Common Compiler Flags
+---------------------
+
+This file contains a single function to configure platform agnostic compiler
+flags like optimization levels, warnings, and features. For platform specific
+flags look to each of the ``cmake/<platform>.cmake`` files.
+
+]=======================================================================]
+#Generator Expression Helpers
+set( IS_CLANG "$<CXX_COMPILER_ID:Clang>" )
+set( IS_APPLECLANG "$<CXX_COMPILER_ID:AppleClang>" )
+set( IS_GNU "$<CXX_COMPILER_ID:GNU>" )
+set( IS_MSVC "$<CXX_COMPILER_ID:MSVC>" )
+set( NOT_MSVC "$<NOT:$<CXX_COMPILER_ID:MSVC>>" )
+
+set( GNU_LT_V8 "$<VERSION_LESS:$<CXX_COMPILER_VERSION>,8>" )
+set( GNU_GE_V9 "$<VERSION_GREATER_EQUAL:$<CXX_COMPILER_VERSION>,9>" )
+set( GNU_GT_V11 "$<VERSION_GREATER_EQUAL:$<CXX_COMPILER_VERSION>,11>" )
+set( GNU_LT_V11 "$<VERSION_LESS:$<CXX_COMPILER_VERSION>,11>" )
+set( GNU_GE_V12 "$<VERSION_GREATER_EQUAL:$<CXX_COMPILER_VERSION>,12>" )
+
+set( HOT_RELOAD-UNSET "$<STREQUAL:${GODOT_USE_HOT_RELOAD},>")
+
+set( DISABLE_EXCEPTIONS "$<BOOL:${GODOT_DISABLE_EXCEPTIONS}>")
+
+
+function( common_compiler_flags TARGET_NAME )
+    set( IS_RELEASE "$<STREQUAL:${TARGET_NAME},template_release>")
+    set( DEBUG_FEATURES "$<OR:$<STREQUAL:${TARGET_NAME},template_debug>,$<STREQUAL:${TARGET_NAME},editor>>" )
+    set( HOT_RELOAD "$<IF:${HOT_RELOAD-UNSET},$<NOT:${IS_RELEASE}>,$<BOOL:${GODOT_USE_HOT_RELOAD}>>" )
+    set( DEBUG_SYMBOLS "$<BOOL:${GODOT_DEBUG_SYMBOLS}>" )
+
+    target_compile_features(${TARGET_NAME}
+            PUBLIC
+            cxx_std_17
+    )
+
+    # These compiler options reflect what is in godot/SConstruct.
+    target_compile_options( ${TARGET_NAME}
+        PUBLIC
```

Comment:
Why is because a lot of things were changed, there was no propagation at all previously, and setting it to public seemed expedient at the time to propagate some flags that were necessary, and I do not have a great mental model for what should be propagated and what should not yet.

I Was indeed hoping that someone like yourself would notify me if there was a problem and we could resolve it, thereby  solving it using reality rather than some made up idealistic notion I might have.

So there's the why, what needs to change? what options can be private that dont effect consumers?

The only thing I see is that you are using `-Wpedantic` and that picks up the anonymous struct business. Unless the godot-cpp source changes, and probably engine too, that will always be a problem for people consuming the headers.

Reply:
@kelteseth I've been testing, I think I can make almost all the flags PRIVATE. I'll subit a PR privatising as much as I can.

Reply:
Nice 🚀